examples/nat_scapy_tests.py

```diff
@@ -0,0 +1,265 @@
+
+# Templates for iptables rules
+iptables_rule_fmt = "INPUT -p tcp --sport %d --dport %d -j DROP"
+iptables_add_rule_fmt = "iptables -A %s"
+iptables_remove_rule_fmt = "iptables -D %s"
+
+# Data to be sent through the NAT
+server_data = "SERVER DATA"
+client_data = "CLIENT DATA"
+
+def run_server(serverport=12012, natport=35001):
+    "This should be run on the external host."
+
+    # Set up iptables rule to allow us exclusive access on our port
+    iptables_rule = iptables_rule_fmt % (natport, serverport)
+    print "  server> $ %s" % (iptables_add_rule_fmt % iptables_rule)
+    subprocess.check_call(iptables_add_rule_fmt % iptables_rule, shell=True)
+
+    # We use the same IP layer and filter each time
+    ip = IP(src=serverhost, dst=nathost)
+    filter = "tcp and host %s and port %d" % (nathost, serverport)
+
+    # Wait for the connection to be initialised from the client
+    print "Waiting for connection from client"
+    syn = sniff(count=1, filter=filter)[0]
+    # send Syn+Ack
+    tcp_synack = TCP(sport=serverport, dport=natport, flags="SA", options=[('MSS', 1460)])
+    print "Sending Syn+Ack"
+    send(ip/tcp_synack)
+
+    # Receive data and check if it is correct
+    print "Waiting for data from client"
+    datap = sniff(count=2, filter=filter)[1]
+    print "Received:"
+    datap.show()
+    data = str(datap[TCP].payload)
+    if (data == client_data):
+        print "Correct data received from the client."
+    else:
+        print "WARNING: incorrect data received from the client: '%s'" % data
+
+    # Wait for connection to be dropped
+    print "Sleeping to wait for the connection to be dropped by the NAT (reliant on tcp_inactivity_timeout=%ds)" % tcp_inactivity_timeout
+    time.sleep(tcp_inactivity_timeout + 2)
+
+    # Try sending data - should fail
+    print "Trying to send a packet to the client - should fail"
+    send(ip/TCP(sport=serverport, dport=natport, flags="", options=[('MSS', 1460)])/"FAIL PLEASE")
+
+    # Remove iptables rule
+    print "  server> $ %s" % (iptables_remove_rule_fmt % iptables_rule)
+    subprocess.check_call(iptables_remove_rule_fmt % iptables_rule, shell=True)
+
+    if (data == client_data):
+        return 0
+    else:
+        return 1
+
+def run_client(serverport=12012, clientport=12011):
+    "This should be run on the internal host."
+
+    # Set up iptables rule to allow us exclusive access on our port
+    iptables_rule = iptables_rule_fmt % (serverport, clientport)
+    print "  client> $ %s" % (iptables_add_rule_fmt % iptables_rule)
+    subprocess.check_call(iptables_add_rule_fmt % iptables_rule, shell=True)
+
+    # We use the same IP layer and filter each time
+    ip = IP(src=clienthost, dst=serverhost)
+    filter = "tcp and host %s and port %d" % (serverhost, clientport)
+
+    # Establish connection to the server
+    tcp_syn = TCP(sport=clientport, dport=serverport, flags="S", options=[('MSS', 1460)])
+    print "Sending syn"
+    synack = sr1(ip/tcp_syn)[0]
```

Comment:
It would be prudent here to check that the Syn and Ack flags in the received packet are indeed set.